### ISSUE TITLE
ctx.rule.attr.srcs empty for go_test in aspect

### ISSUE THREAD

**Issue by chaoren** [bazelbuild/rules_go#1251]
```
$ cat WORKSPACE
git_repository(
    name = "io_bazel_rules_go",
    remote = "https://github.com/bazelbuild/rules_go.git",
    tag = "0.8.1",
)

load("@io_bazel_rules_go//go:def.bzl", "go_rules_dependencies", "go_register_toolchains")
go_rules_dependencies()
go_register_toolchains()

$ cat BUILD
load("@io_bazel_rules_go//go:def.bzl", "go_prefix", "go_library", "go_test")

go_prefix("foo")

go_library(
    name = "foo",
    srcs = ["foo.go"],
)

go_test(
    name = "foo_test",
    srcs = ["foo_test.go"],
    library = ":foo",
)

$ cat aspect.bzl
def _aspect_impl(target, ctx):
    print(target, ctx.rule.attr.srcs)
    return []


srcs_aspect = aspect(
    implementation = _aspect_impl,
)

$ bazel build :foo :foo_test --aspects=//:aspect.bzl%srcs_aspect
DEBUG: /path/to/aspect.bzl:2:5: <target //:foo> [<input file target //:foo.go>]
DEBUG: /path/to/aspect.bzl:2:5: <target //:foo_test> []
INFO: Analysed 2 targets (0 packages loaded).
INFO: Found 2 targets...
INFO: Elapsed time: 1.219s, Critical Path: 0.00s
INFO: Build completed successfully, 1 total action
```

This breaks bazel/go/intellij integration. See https://github.com/bazelbuild/intellij/issues/214.

**Comment by chaoren** [comment]
Issue did not exist in 0.5.5, so this is a regression.

**Comment by ianthehat** [comment]
I don't think this is a rules_go issue...
I'm not sure what you are hoping for here, it's allowed to have an empty source. In fact, I think it's now guaranteed that it will *always* have no declared source (it will however have generated source, specifically the test main, but that wont be visible in the attributes)
the `go_test` you see as the rules_go api is the macro, the `go_test` you see in the aspect is an internal implementation detail.

**Comment by brendandouglas** [comment]
We'll need some way to access the target's sources (both generated and non-generated). This is needed from all language rules for IDE integration. 

If it's not available via 'ctx.rule.attr.srcs', we'll need to look at exposing it via a go-specific skylark provider. I'd definitely recommend the more common 'ctx.rule.attr.srcs' though.

From a user's perspective, the fact that go_test is implemented via a macro is the implementation detail -- all they see is go_test(srcs = [],...).

**Comment by brendandouglas** [comment]
Any particular reason why 'go_test_macro' wouldn't pass the original srcs on to go_test?

go_test still has a 'srcs' attribute, but it appears entirely unused (both in the callers and in the actual go_test_impl implementation).

Considering this breaks IDE integration, any objections to a quick fix passing on srcs?

**Comment by ianthehat** [comment]
You want the [GoSource](/bazelbuild/rules_go/blob/master/go/providers.rst#gosource) provider.

From a users perspective, they only see the macro, and that's my point. Aspects only get to see the underlying rules, and even then only after select processing, so they are never going to give you the same information that is in the build files, a significant fraction of the information has been thrown away well before the aspect triggers. We don't even have the same structure of rules in different modes (cgo causes a totally different set of rules to instantiate).
Also you are examining something that is not the public API of rules_go, we make no compatibility promises about the information you are looking at if you look at the rule attributes.

**Comment by ianthehat** [comment]
go_test (the underlying rule, not the macro) does not want or need those sources. In fact it makes no sense for it to have them, they are not even in the same package as it's actual source.

go_test represents the test binary itself, not the package under test. The srcs are passed to a proper go library rule that is then linked into the test binary.
I plan to delete that attribute in the near future, it's a left over from an old mechanism that no longer exists.

**Comment by brendandouglas** [comment]
Thanks for the link to the provider, we should be able to use that in the short term.

Longer term, are there any plans for a public Skylark API (or any plans to move some of the rules_go code internally)? IDE integration is an important use case here, and will eventually require a stable API.

**Comment by ianthehat** [comment]
The GoSource provider is intended to be a stable public API (although before we hit 1.0 we reserve the right to change just about anything until we think we have a good system).
The structure of rules that the aspect gets to see is not (and will probably never be) a stable API.
The plan for all tools in the longer term is to provide an build system neutral abstraction layer that can report go packages and what their contents are. Tools written on top of that API will be able to work in a GOPATH environment just as easily as a Bazel one. 
The Bazel implementation of that abstraction layer may use aspects to do it's work, or it might not, but it will be allowed to use a non public API because it will be part of the rules_go repository and maintained by us. We are highly unlikely to ever constrain ourselves to not changing what the world looks like to an aspect, it would prevent us from making progress.

In the short/medium term if you would like to talk about what information you need, and what it's format should be, and maybe some tests, I would be happy to add skylark to rules_go to produce that information, and keep it working even when the underlying data changes.

If you look at the stuff I did in the go_path rules, it probably gets you the information you want/need, something based on that would be easy to add.

**Comment by brendandouglas** [comment]
Thanks Ian, that sounds like a good plan. I expect we'll be able to use the GoSource provider for now. Should be fine to close this -- I'll raise a separate issue if we need some more information exposed in skylark.
